internal/server/server.go

```diff
@@ -0,0 +1,240 @@
+		return
+	}
+
+	isCorrectPassword, err := crypto.IsCorrectPassword(account, password)
+	if err != nil {
+		log.Println("Salt stored in database is incorrectly encoded")
+		c.AbortWithStatus(http.StatusInternalServerError)
+		return
+	}
+
+	if !isCorrectPassword {
+		log.Println("Bad password")
+		c.AbortWithStatus(http.StatusUnauthorized)
+		return
+	}
+
+	// both session tuples should only be accessed by this user, so we won't bother
+	// with a transaction
+	authenticatedSession, err := env.db.CreateSession(account.AccountId, csrfToken, now)
+	if err != nil {
+		log.Printf("Error creating session: %v\n", err)
+		c.AbortWithStatus(http.StatusUnauthorized)
+		return
+	}
+
+	err = env.db.DeleteSession(sessionToken)
+	if err != nil {
+		log.Printf("Error deleting old session: %v\n", err)
+		c.AbortWithStatus(http.StatusUnauthorized)
+		return
+	}
+
+	sessionCookie := http.Cookie{
+		Name:     "session_token",
+		Value:    authenticatedSession.SessionToken,
+		Path:     "/",
+		Secure:   true,
+		HttpOnly: true,
+		SameSite: http.SameSiteLaxMode,
+	}
+
+	http.SetCookie(c.Writer, &sessionCookie)
+	c.AbortWithStatus(http.StatusOK)
+}
+
+func (env *Env) logout(c *gin.Context) {
+	// TODO: move logged in check to its own function
+	now := time.Now()
+	csrfToken := c.Request.Header.Get("CSRF")
+
+	sessionToken, err := c.Cookie("session_token")
+	// should look into deduplicating this
+	// should also look into a logging library w/ levels (debug, info, warning, etc)
+	if err != nil {
+		log.Println("Missing session token")
+		c.AbortWithStatus(http.StatusUnauthorized)
+		return
+	}
+
+	session, err := env.db.FetchSession(sessionToken)
+	if err != nil {
+		log.Printf("Error fetching session: %v\n", err)
+		c.AbortWithStatus(http.StatusUnauthorized)
+		return
+	}
+
+	if session.Expired(now) {
+		log.Println("Session expired")
+		c.AbortWithStatus(http.StatusOK)
+		return
+	}
+
+	if !crypto.IsSessionOwner(session, csrfToken) {
+		log.Println("Not owner of session")
+		c.AbortWithStatus(http.StatusUnauthorized)
+		return
+	}
+
+	err = env.db.DeleteSession(sessionToken)
+	if err != nil {
+		log.Printf("Error deleting session: %v\n", err)
+	}
+
+	c.AbortWithStatus(http.StatusOK)
+}
+
+func SetupRouter(env *Env) *gin.Engine {
+	router := gin.Default()
+	router.LoadHTMLGlob(env.webroot + "index.html")
+	router.GET("/ping", env.ping)
+	router.GET("/index.html", env.serveIndex)
+	router.GET("/", env.serveIndex)
+	router.POST("/api/login", env.login)
+	router.POST("/api/logout", env.logout)
+
+	router.Static("/static", env.webroot+"static")
+	rootLevelFiles := []string{
+		"asset-manifest.json",
+		"favicon.ico",
+		"logo192.png",
+		"logo512.png",
+		"manifest.json",
+		"robots.txt",
+	}
+
+	for _, file := range rootLevelFiles {
+		router.StaticFile(fmt.Sprintf("/%s", file), fmt.Sprintf("./web/%s", file))
+	}
+
+	return router
+}
+
+func Run(env *Env) error {
+	go func() {
+		tick := time.NewTicker(time.Minute)
+		for {
+			now := <-tick.C
```

Comment:
Pass context to Run method and do a select on this and context.Done() channel - this way this goroutine will be cancelable.